src/library/scala/collection/immutable/OrderedMap.scala

```diff
@@ -0,0 +1,426 @@
+      case Some((o, _)) if orderedBy == OrderBy.Insertion =>
+        new OrderedMap(
+          ordering.incl(o, key),
+          mapping.updated[(Int, V1)](key, (o, value)),
+          o,
+          orderedBy)
+      case Some((o, _)) =>
+        val o1 = if (ordinal == Int.MaxValue) Int.MinValue else ordinal + 1
+        new OrderedMap(
+          ordering.excl(o).incl(o1, key),
+          mapping.updated[(Int, V1)](key, (o1, value)),
+          o1,
+          orderedBy)
+      case None ⇒
+        val o1 = ordinal + 1
+        new OrderedMap(
+          ordering.incl(o1, key),
+          mapping.updated[(Int, V1)](key, (o1, value)),
+          o1,
+          orderedBy)
+    }
+  }
+
+  def remove(key: K): OrderedMap[K, V] = {
+    mapping.get(key) match {
+      case Some((o, _)) ⇒
+        new OrderedMap(
+          ordering.excl(o),
+          mapping.remove(key),
+          ordinal,
+          orderedBy)
+      case None ⇒
+        this
+    }
+  }
+
+  def refresh(key: K): OrderedMap[K, V] = {
+    mapping.get(key) match {
+      case Some((o, _)) ⇒
+        val o1 = ordinal + 1
+        new OrderedMap(
+          ordering.excl(o).incl(o1, key),
+          mapping,
+          o1,
+          orderedBy)
+      case None ⇒
+        this
+    }
+  }
+
+  override def contains(key: K): Boolean = mapping.contains(key)
+
+  override def head: (K, V) = binding(ordering.head)
+
+  override def headOption = ordering.headOption.map(binding)
+
+  override def last: (K, V) = binding(ordering.last)
+
+  override def lastOption: Option[(K, V)] = ordering.lastOption.map(binding)
+
+  override def tail: OrderedMap[K, V] = {
+    val (head, tail) = ordering.headTail
+    new OrderedMap(tail, mapping.remove(head), ordinal, orderedBy)
+  }
+
+  override def init: OrderedMap[K, V] = {
+    val (init, last) = ordering.initLast
+    new OrderedMap(init, mapping.remove(last), ordinal, orderedBy)
+  }
+
+  def get(key: K): Option[V] = mapping.get(key).map(value)
+
+  def iterator: Iterator[(K, V)] = new AbstractIterator[(K, V)] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): (K, V) = binding(iter.next())
+  }
+
+  override def keysIterator: Iterator[K] = new AbstractIterator[K] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): K = iter.next()
+  }
+
+  override def valuesIterator: Iterator[V] = new AbstractIterator[V] {
+    private[this] val iter = ordering.iterator
+
+    override def hasNext: Boolean = iter.hasNext
+
+    override def next(): V = value(binding(iter.next()))
+  }
+
+  override def concat[V1 >: V](suffix: IterableOnce[(K, V1)]): OrderedMap[K, V1] = {
+    var result: OrderedMap[K, V1] = this
+    val iter = suffix.iterator
+    while (iter.hasNext) {
+      val (k, v) = iter.next()
+      result = result.updated(k, v)
+    }
+    result
+  }
```

Comment:
if `suffix` is a mutable/immutable HashMap, `concat` could avoid rehashing keys when inserting into the OrderedMap's HashMap.

Same with `removedAll` if the argument is a mutable/immutable HashSet

Reply:
@joshlemer I'm a bit uncertain as to how that would be accomplished so I'll take a look at the implementations in HashMap/HashSet for guidance.

Reply:
This is the PR to check https://github.com/scala/scala/pull/7520. Maybe it's easier to just wait until that is merged so that you get the private[collection] methods `immutable.HashMap/HashSet#foreachWithHash` and `mutable.HashMap/HashSet#nodeIterator`

specifically, 

https://github.com/scala/scala/pull/7520/files#diff-5be35908b1ec9bbb90c66342b71dd9afR90

shows how both can be used.

Reply:
@joshlemer Do you still feel that this issue needs addressing considering the latest set of changes (00b7f1e)?

Reply:
Very cool! 👏 💐

Reply:
Though, this particular comment is more about extracting the hashes out of `suffix` in the case that it happens to be a `HashMap` That way, you can insert the keys into your builder without re-hashing. 

This could be considered a stretch goal though, I don't think it needs to block this PR, maybe we should make a separate backlog-milestone issue for it.